src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -0,0 +1,208 @@
+import com.snowflake.kafka.connector.Utils;
+import com.snowflake.kafka.connector.internal.KCLogger;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import net.snowflake.ingest.streaming.SnowflakeStreamingIngestClient;
+import net.snowflake.ingest.streaming.SnowflakeStreamingIngestClientFactory;
+import net.snowflake.ingest.utils.SFException;
+import org.apache.kafka.connect.errors.ConnectException;
+
+/** Singleton that provides the streaming client */
+public class StreamingClientProvider {
+  public static final StreamingClientProvider streamingClientProvider =
+      new StreamingClientProvider();
+
+  /**
+   * Checks if the client is valid by doing a null check and ensuring it is open
+   *
+   * @param client The client to validate
+   * @return If the client is not null and open
+   */
+  public static boolean isClientValid(SnowflakeStreamingIngestClient client) {
+    return client != null && !client.isClosed();
+  }
+
+  /**
+   * ONLY FOR TESTING - to get a provider with injected properties
+   *
+   * @param createdClientId the number of times a client has been created
+   * @param connectorConfig the connector config
+   * @param client the injected streaming client
+   * @return a provider with the injected properties
+   */
+  @VisibleForTesting
+  public static StreamingClientProvider injectStreamingClientProviderForTests(
+      int createdClientId,
+      Map<String, String> connectorConfig,
+      SnowflakeStreamingIngestClient client) {
+    return new StreamingClientProvider(createdClientId, connectorConfig, client);
+  }
+
+  private static final KCLogger LOGGER = new KCLogger(Utils.class.getName());
+  private static final String STREAMING_CLIENT_PREFIX_NAME = "KC_CLIENT_";
+
+  private int createdClientId;
+  private Map<String, String> connectorConfig;
+  private SnowflakeStreamingIngestClient streamingIngestClient;
+
+  // private constructor for singleton
+  private StreamingClientProvider() {
+    this.createdClientId = 0;
+    this.connectorConfig = new HashMap<>();
+  }
+
+  @VisibleForTesting
+  // ONLY FOR TESTING - private constructor to inject properties for testing
+  private StreamingClientProvider(
+      int createdClientId,
+      Map<String, String> connectorConfig,
+      SnowflakeStreamingIngestClient client) {
+    this.createdClientId = createdClientId;
+    this.connectorConfig = connectorConfig;
+    this.streamingIngestClient = client;
+  }
+
+  /**
+   * Creates a new streaming client, will replace the existing one if necessary
+   *
+   * @param connectorConfig The connector config to define the client
+   */
+  public void createClient(Map<String, String> connectorConfig) {
+    // replace previous connector config and client if applicable
+    if (!this.connectorConfig.isEmpty()) {
+      LOGGER.warn("Overriding previous connector config");
+    }
```

Comment:
Is this a valid code path? If yes, then we should use INFO?

Reply:
It is not a valid code path, but I removed this warning as part of the close client validation comment earlier